l10n_fi_invoice_numbers/models/account_move.py

```diff
@@ -0,0 +1,63 @@
+##############################################################################
+#
+#    Author: Avoin.Systems
+#    Copyright 2015-2019 Avoin.Systems
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from odoo import models, fields, api
+# noinspection PyProtectedMember
+from odoo.tools.translate import _
+import re
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    @api.depends('name', 'state')
+    def _compute_ref_number(self):
+        for invoice in self:
+            if invoice.name:
+                invoice_number = re.sub(r'\D', '', invoice.name)
+                checksum = sum((7, 3, 1)[idx % 3] * int(val)
+                               for idx, val in enumerate(invoice_number[::-1]))
+                invoice.ref_number = invoice_number + str((10 - (checksum % 10)) % 10)
```

Comment:
This calculation already seems to be done in the `l10n_fi_payment_reference` module: https://github.com/OCA/l10n-finland/blob/11.0/l10n_fi_payment_reference/models/account_invoice.py#L53

This module should either:
- Use that module so the logic isn't duplicated unnecessarily
- Have a comment or other explanation of why it needs to re-calculate the finnish reference

As part of this you should consider whether it is necessary to skip RF-type references for some reason. This module does not support them, but the `l10n_fi_payment_reference` module does. Is that significant?

Reply:
It is possible this module is no longer required: the invoice template could just directly use the `payment_reference` field and separate the checksum-digit from it as necessary.

Reply:
And not even the `payment_reference` field, but the `invoice_payment_ref` that Odoo has introduced https://github.com/odoo/odoo/blob/31890c4e676f7f4f65221ca66bfd66cc0ceedecc/addons/account/models/account_move.py#L200